chore: Document the three OpenTelemetry setups in the migration guide (#23149)
getsentry/sentry-javascript: MIGRATION.md

Reviewer: l: Should we add this so that users/agents can quickly spot if this concerns them? 

```suggestion
Affected SDKs: Server-side SDKs (`@sentry/node` and all dependents). **If Sentry is used with a custom OpenTelemetry SDK Setup**
```

Author: No this affects v11 as a whole, not just user's with custom otel setups.

This means you can run your own OpenTelemetry setup cleanly alongside Sentry without having Sentry spans leak into your pipeline anymore. Your OpenTelemetry setup will no longer be required to use Sentry components for exporting, context management and trace propagation.

With this, we also heavily reduced our OpenTelemetry dependencies, with `@opentelemetry/api` being the only remaining package we abide by. These changes also mean `@sentry/node-core` no longer serves any purpose and was [merged back into `@sentry/node`](#sentrynode-core-was-merged-back-into-sentrynode).

For most users, day-to-day tracing is **unchanged**.

#### Choosing an OpenTelemetry setup

There are three ways to run the two together, and which one you want depends on who should own spans. This is controlled by the existing `skipOpenTelemetrySetup` option, whose default was flipped in v11: it is now `true` for most server SDKs (including `@sentry/node`, `@sentry/bun`, the serverless SDKs and `@sentry/cloudflare`) and `false` for `@sentry/nextjs` and `@sentry/sveltekit`.

> ##### 1. Sentry only, OpenTelemetry ignored

Reviewer 2: ```suggestion
##### 1. Sentry only
```

I would just call this "Sentry only" or if you want to have OTel in the title "Sentry only, no OpenTelemetry"

Author: Good call

> This is the default for every server-side SDK except `@sentry/nextjs` and `@sentry/sveltekit`. Nothing to configure:

Reviewer 2: this section is just about the general use case "Sentry only" right? in that case people don't really care about anything OTel I think, so imo we don't even have to mention that some SDKs set up a tracer provider, this is all under the hood and "just works" so users probably don't care

Author: Fair, simplified it to not mention OTel at all.

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
});

> ```

Reviewer 2: since this is just the default do we even need an example? I would just say something along the lines of "if you use sentry only and don't want to use OTel alongside it just set up your SDK normally you are good" and then maybe the note at the bottom to direct them to step 2 if they want to create spans through `@opentelemetry/api`

Author: I think it's clearer this way especially as you can compare it quickly at a glance with the other two cases.

The SDK emits native Sentry spans and never registers an OpenTelemetry tracer provider or propagator. Scope isolation uses `AsyncLocalStorage` directly.

Spans created through `@opentelemetry/api` are **ignored**. If a library you depend on emits OpenTelemetry spans and you want them in Sentry, use setup 2.

##### 2. OpenTelemetry-compatible mode, everything goes to Sentry

Set `skipOpenTelemetrySetup: false`:

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
  skipOpenTelemetrySetup: false,
});
```

> Sentry registers a minimal tracer provider, context manager and propagator. Just enough OpenTelemetry to pick up spans created through `@opentelemetry/api`, which become native Sentry spans.

Reviewer 2: "a minimal OTel tracer provider" or maybe "OTel compatible"?

Author: Updated to OpenTelemetry-compatible.

This means you can run your own OpenTelemetry setup cleanly alongside Sentry without having Sentry spans leak into your pipeline anymore. Your OpenTelemetry setup will no longer be required to use Sentry components for exporting, context management and trace propagation.

With this, we also heavily reduced our OpenTelemetry dependencies, with `@opentelemetry/api` being the only remaining package we abide by. These changes also mean `@sentry/node-core` no longer serves any purpose and was [merged back into `@sentry/node`](#sentrynode-core-was-merged-back-into-sentrynode).

For most users, day-to-day tracing is **unchanged**.

#### Choosing an OpenTelemetry setup

There are three ways to run the two together, and which one you want depends on who should own spans. This is controlled by the existing `skipOpenTelemetrySetup` option, whose default was flipped in v11: it is now `true` for most server SDKs (including `@sentry/node`, `@sentry/bun`, the serverless SDKs and `@sentry/cloudflare`) and `false` for `@sentry/nextjs` and `@sentry/sveltekit`.

##### 1. Sentry only

> The default, and what most users want. Tracing works out of the box:

Reviewer: l: I think we should use "you" or "users" consistently. So far I think we've been using "you"

```suggestion
The default, and what you most likely want. Tracing works out of the box:
```

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
});
```

If a library you depend on emits its own OpenTelemetry spans and you want those in Sentry too, use setup 2.

##### 2. OpenTelemetry-compatible mode, everything goes to Sentry

Set `skipOpenTelemetrySetup: false`:

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
  skipOpenTelemetrySetup: false,
});
```

Sentry registers a minimal tracer provider, context manager and propagator. Just enough OpenTelemetry to pick up spans created through `@opentelemetry/api`, which become native Sentry spans.

This is what `@sentry/nextjs` and `@sentry/sveltekit` do by default, because those frameworks emit OpenTelemetry spans that would otherwise be lost.

> Everything goes to Sentry and only to Sentry. This is not a general OpenTelemetry pipeline: there is no exporter, no OTLP output, and no way to fan spans out to another backend. Sentry also refuses to register its provider if you already registered one of your own, logging a warning instead. If you want a real OpenTelemetry pipeline, use setup 3.

Reviewer: very technically, and probably fine to omit here, users could still roll their own transport and fan out to other tracing backends. But I think we discussed this already so probably not worth mentioning here.

This means you can run your own OpenTelemetry setup cleanly alongside Sentry without having Sentry spans leak into your pipeline anymore. Your OpenTelemetry setup will no longer be required to use Sentry components for exporting, context management and trace propagation.

> With this, we also heavily reduced our OpenTelemetry dependencies, with `@opentelemetry/api` being the only remaining package we abide by. These changes also mean `@sentry/node-core` no longer serves any purpose and was [merged back into `@sentry/node`](#sentrynode-core-was-merged-back-into-sentrynode).

Reviewer 3: ```suggestion
With this, we also heavily reduced our OpenTelemetry dependencies, with `@opentelemetry/api` being the only one remaining. These changes also mean `@sentry/node-core` no longer serves any purpose and was [merged back into `@sentry/node`](#sentrynode-core-was-merged-back-into-sentrynode).
```

Reviewer 3: I cannot under any circumstance accept this "abide by"

Reviewer 2: I called this out as clanker slop the first time I read this, but Andrei swore it's his own words

Author: I wrote this... I won't entertain this accusation. I will simply not abide by it.

Reviewer 3: Then it's two people calling this out already, it must go

> For most users, day-to-day tracing is **unchanged**.

Reviewer: l: wdyt? 

```suggestion
If you only use the Sentry SDK, day-to-day tracing remains **unchanged**.
```

#### Choosing an OpenTelemetry setup

> There are three ways to run the two together, and which one you want depends on who should own spans. This is controlled by the existing `skipOpenTelemetrySetup` option, whose default was flipped in v11: it is now `true` for most server SDKs (including `@sentry/node`, `@sentry/bun`, the serverless SDKs and `@sentry/cloudflare`) and `false` for `@sentry/nextjs` and `@sentry/sveltekit`.

Reviewer: ```suggestion
There are three ways to run the Sentry and OTel SDKs together, and which one you want depends on who should own spans. This is controlled by the existing `skipOpenTelemetrySetup` option, whose default was flipped in v11: it is now `true` for most server SDKs (including `@sentry/node`, `@sentry/bun`, the serverless SDKs and `@sentry/cloudflare`) and `false` for `@sentry/nextjs` and `@sentry/sveltekit`.
```

##### 1. Sentry only, OpenTelemetry ignored

This is the default for every server-side SDK except `@sentry/nextjs` and `@sentry/sveltekit`. Nothing to configure:

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
});
```

The SDK emits native Sentry spans and never registers an OpenTelemetry tracer provider or propagator. Scope isolation uses `AsyncLocalStorage` directly.

Spans created through `@opentelemetry/api` are **ignored**. If a library you depend on emits OpenTelemetry spans and you want them in Sentry, use setup 2.

##### 2. OpenTelemetry-compatible mode, everything goes to Sentry

Set `skipOpenTelemetrySetup: false`:

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
  skipOpenTelemetrySetup: false,
});
```

Sentry registers a minimal tracer provider, context manager and propagator. Just enough OpenTelemetry to pick up spans created through `@opentelemetry/api`, which become native Sentry spans.

> This is what `@sentry/nextjs` and `@sentry/sveltekit` do by default, because those frameworks emit OpenTelemetry spans that would otherwise be lost.

Reviewer 2: do we need a user to know that? seems like implementation detail to me

Author: Agreed, dropped.

Everything goes to Sentry and only to Sentry. This is not a general OpenTelemetry pipeline: there is no exporter, no OTLP output, and no way to fan spans out to another backend. Sentry also refuses to register its provider if you already registered one of your own, logging a warning instead. If you want a real OpenTelemetry pipeline, use setup 3.

##### 3. Your own OpenTelemetry, Sentry linked to it

> Turn Sentry tracing off, own the OpenTelemetry setup yourself, and add `otlpIntegration()`:

Reviewer 2: ```suggestion
Turn Sentry tracing off, use your own OpenTelemetry setup, and add the Sentry `otlpIntegration()`:
```

Sentry.init({
  dsn: '__DSN__',
  // no tracesSampleRate: OpenTelemetry owns spans, Sentry owns errors and logs
  integrations: [Sentry.otlpIntegration()],
});
```

> OpenTelemetry owns spans end to end. Sentry captures errors and logs, and `otlpIntegration()` attaches them to whatever OpenTelemetry span is active so they land on the same trace. `getOtlpTracesEndpoint()` turns your DSN into the URL and auth headers for Sentry's OTLP endpoint, so you can point your own exporter at Sentry, at your own collector, or at both.

Reviewer 2: ```suggestion
OpenTelemetry owns spans end to end. Sentry captures errors and logs, and the Sentry `otlpIntegration()` attaches them to the active OpenTelemetry span so all your telemetry is connected in one trace. `getOtlpTracesEndpoint()` turns your DSN into the URL and auth headers for Sentry's OTLP endpoint, so you can point your own exporter at Sentry, at your own collector, or at both.
```

Sentry does not touch your pipeline: no exporter, no span processor, no tracer provider, and outgoing trace propagation is left to your propagator. See [Connecting Sentry to your OpenTelemetry traces](#connecting-sentry-to-your-opentelemetry-traces) for the details, including what changed if you used the v10 integration.

##### Avoiding duplicate spans

> Leaving `tracesSampleRate` unset is what keeps setup 3 clean. Sentry instruments many of the same libraries OpenTelemetry does (Express, Postgres, Redis, Prisma, Kafka and so on), so enabling Sentry tracing on top of your own instrumentation gives you two spans for every operation. With tracing off, Sentry's instrumentation stays installed and keeps doing request isolation, but emits no spans, so there is nothing to collide.

Reviewer 2: ```suggestion
Sentry instruments many of the same libraries OpenTelemetry does (Express, Postgres, Redis, Prisma, Kafka and so on), so enabling Sentry tracing on top of your own instrumentation gives you two spans for every operation. Leave `traceSampleRate` in your `Sentry.init` unset to avoid duplicate spans. With tracing off, Sentry's instrumentation stays installed and keeps isolating requests, but emits no spans.
```

> Sentry.init({

Reviewer 2: should we expand this example to also include the rest of the setup with how to configure the exporter with the OTLP endpoint?

Author: `  spanProcessors: [new BatchSpanProcessor(new OTLPTraceExporter(Sentry.getOtlpTracesEndpoint('__DSN__')))],` you mean this?

It's part of the example.

Reviewer 2: yes right, I missed that sorry

Sentry does not touch your pipeline: no exporter, no span processor, no tracer provider, and outgoing trace propagation is left to your propagator. See [Connecting Sentry to your OpenTelemetry traces](#connecting-sentry-to-your-opentelemetry-traces) for the details, including what changed if you used the v10 integration.

##### Avoiding duplicate spans

Leaving `tracesSampleRate` unset is what keeps setup 3 clean. Sentry instruments many of the same libraries OpenTelemetry does (Express, Postgres, Redis, Prisma, Kafka and so on), so enabling Sentry tracing on top of your own instrumentation gives you two spans for every operation. With tracing off, Sentry's instrumentation stays installed and keeps doing request isolation, but emits no spans, so there is nothing to collide.

Note that this changed since v10, where setting `skipOpenTelemetrySetup: true` also turned Sentry's HTTP and fetch spans off by default. Sentry now emits those whenever tracing is enabled, regardless of `skipOpenTelemetrySetup`.

> If you do want Sentry spans alongside your own, drop the integrations that overlap. HTTP and fetch are the exception: keep those two and turn off only their spans, because `httpIntegration` also provides request isolation, request data and session tracking.

Reviewer 2: ```suggestion
If you do want Sentry spans alongside your own, keep `traceSampleRate` set and drop the integrations that overlap. HTTP and fetch are the exception: turn off only their spans, because `httpIntegration` also provides request isolation, request data and session tracking:
```

```js
Sentry.init({
  dsn: '__DSN__',
  tracesSampleRate: 1.0,
  integrations: integrations => [
    // your own OpenTelemetry instrumentation already covers these
    ...integrations.filter(integration => integration.name !== 'Postgres'),
    Sentry.httpIntegration({ spans: false }),
    Sentry.nativeNodeFetchIntegration({ spans: false }),
  ],
});
```

> ##### The v10 bridge is gone

Reviewer 2: can we rename this? sounds quite sloppy to me. we could just reuse the name we used so far for this "Custom OTel setup" or something like this?

Author: Yes, renamed to "Migrate custom OpenTelemetry setups"
